components/push/src/communications.rs

```diff
@@ -697,5 +694,228 @@ mod test {
+                ..Default::default()
+            };
+            // We first subscribe to get a UAID
+            let body = json!({
+                "uaid": DUMMY_UAID,
+                "channelID": DUMMY_CHID,
+                "endpoint": "https://example.com/update",
+                "senderid": SENDER_ID,
+                "secret": SECRET,
+            })
+            .to_string();
+            let ap_mock = mock(
+                "POST",
+                format!("/v1/test/{}/registration", SENDER_ID).as_ref(),
+            )
+            .with_status(200)
+            .with_header("content-type", "application/json")
+            .with_body(body)
+            .create();
+            let mut conn = connect(config, None, None).unwrap();
+            let channel_id = hex::encode(crate::crypto::get_random_bytes(16).unwrap());
+            let response = conn.subscribe(&channel_id, None).unwrap();
+            ap_mock.assert();
+            assert_eq!(response.uaid, conn.uaid.clone().unwrap());
+            // We mock that the server is returning a client error
+            let channel_list_body = json!({
+                "code": status_codes::INTERNAL_SERVER_ERROR,
+                "errno": 999,
+                "error": "",
+                "message": "Unknown Error"
+
+            })
+            .to_string();
+            let channel_list_mock = mock(
+                "GET",
+                format!("/v1/test/{}/registration/{}", SENDER_ID, DUMMY_UAID).as_ref(),
+            )
+            .with_status(status_codes::INTERNAL_SERVER_ERROR as usize)
+            .with_header("content-type", "application/json")
+            .with_body(channel_list_body)
+            .create();
+            // we verify that the call to `verify_connection` didn't error out
+            // and instead is instructing its caller to wipe the UAID from
+            // persisted storage
+            let err = conn.verify_connection(&[channel_id]).unwrap_err();
+            channel_list_mock.assert();
+            assert!(matches!(
+                err.kind(),
+                error::ErrorKind::CommunicationServerError(_)
+            ));
+            // we double check that we did not wipe our uaid
+            assert!(conn.uaid.is_some());
+        }
+
+        // we test that we properly return a `AlreadyRegisteredError` when a client
+        // gets a `CONFLICT` status code
+        {
+            let config = PushConfiguration {
+                http_protocol: Some("http".to_owned()),
+                server_host: server_address().to_string(),
+                sender_id: SENDER_ID.to_owned(),
+                bridge_type: Some("test".to_owned()),
+                registration_id: Some("SomeRegistrationValue".to_owned()),
+                ..Default::default()
+            };
+            // We mock that the server thinks
+            // we already registered!
+            let body = json!({
+                "code": status_codes::CONFLICT,
```

Comment:
Looking at https://autopush.readthedocs.io/en/latest/http.html#error-codes, `409` isn't listed on the error codes....

I'm a little suspicious that the whole `AlreadyRegistered` error might never be hit. I'll take a little time to see if I can confirm that and get rid of the `AlreadyRegistered` error all together, but that's probably better for another clean-up PR